Python/specialize.c

```diff
@@ -0,0 +1,194 @@
+{
+    int cache_offset = 0;
+    int previous_opcode = -1;
+    for (int i = 0; i < len; i++) {
+        uint8_t opcode = _Py_OPCODE(code[i]);
+        if (previous_opcode != EXTENDED_ARG) {
+            oparg_from_instruction_and_update_offset(i, opcode, 0, &cache_offset);
+        }
+        previous_opcode = opcode;
+    }
+    return cache_offset + 1;   // One extra for the count entry
+}
+
+static inline _Py_CODEUNIT *
+first_instruction(SpecializedCacheOrInstruction *quickened)
+{
+    return &quickened[get_cache_count(quickened)].code[0];
+}
+
+/** Insert adaptive instructions and superinstructions.
+ *
+ * Skip instruction preceded by EXTENDED_ARG for adaptive
+ * instructions as those are both very rare and tricky
+ * to handle.
+ */
+static void
+optimize(SpecializedCacheOrInstruction *quickened, int len)
+{
+    _Py_CODEUNIT *instructions = first_instruction(quickened);
+    int cache_offset = 0;
+    int previous_opcode = -1;
+    for(int i = 0; i < len; i++) {
+        int opcode = _Py_OPCODE(instructions[i]);
+        int oparg = _Py_OPARG(instructions[i]);
+        uint8_t adaptive_opcode = adaptive_opcodes[opcode];
+        if (adaptive_opcode && previous_opcode != EXTENDED_ARG) {
+            int new_oparg = oparg_from_instruction_and_update_offset(
+                i, opcode, oparg, &cache_offset
+            );
+            if (new_oparg < 0) {
+                /* Not possible to allocate a cache for this instruction */
+                previous_opcode = opcode;
+                continue;
+            }
+            instructions[i] = _Py_MAKECODEUNIT(adaptive_opcode, new_oparg);
+            previous_opcode = adaptive_opcode;
+            int entries_needed = cache_requirements[opcode];
+            if (entries_needed) {
+                /* Initialize the adpative cache entry */
+                int cache0_offset = cache_offset-entries_needed;
+                SpecializedCacheEntry *cache =
+                    _GetSpecializedCacheEntry(instructions, cache0_offset);
+                cache->adaptive.original_oparg = oparg;
+                cache->adaptive.counter = 0;
+            }
+        }
+        else {
+            /* Super instructions don't use the cache,
+             * so no need to update the offset. */
+            switch (opcode) {
+                /* Insert superinstructions here
+                 E.g.
+                case LOAD_FAST:
+                    if (previous_opcode == LOAD_FAST)
+                        instructions[i-1] = _Py_MAKECODEUNIT(LOAD_FAST__LOAD_FAST, oparg);
+                 */
+            }
+            previous_opcode = opcode;
+        }
+    }
+    assert(cache_offset+1 == get_cache_count(quickened));
+}
+
+int
+_Py_Quicken(PyCodeObject *code) {
+    if (code->co_quickened) {
+        return 0;
+    }
+    Py_ssize_t size = PyBytes_GET_SIZE(code->co_code);
+    int instr_count = (int)(size/sizeof(_Py_CODEUNIT));
+    if (instr_count > MAX_SIZE_TO_QUICKEN) {
```

Comment:
Would it be possible instead to quicken the first 5000 instructions and then exit?  (That would avoid a cliff where a minor change in the code tips it over the no-optimization limit and makes it run much slower.)

Reply:
There's a memory cost to creating a duplicate code array, so it would risk wasting memory unproportionally.